File: app/src/commands/infrastructure/building.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { invoke } from "@tauri-apps/api/core";
import type { BuildingId, Coord, ToggleBuildingRequest } from "@tsukilabs/nil-bindings";

export async function toggleBuilding(coord: Coord, id: BuildingId, enabled: boolean) {
  const req: ToggleBuildingRequest = {
    world: NIL.world.getIdStrict(),
    id,
    coord,
    enabled,
  };

  await invoke("toggle_building", { req });
}

File: crates/nil-payload/src/request/infrastructure/building.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use bon::Builder;
use nil_core::continent::coord::Coord;
use nil_core::infrastructure::building::BuildingId;
use nil_core::world::config::WorldId;
use serde::{Deserialize, Serialize};

#[derive(Builder, Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(feature = "typescript", derive(ts_rs::TS))]
#[cfg_attr(feature = "typescript", ts(export))]
pub struct ToggleBuildingRequest {
  #[builder(start_fn, into)]
  pub world: WorldId,
  #[builder(into)]
  pub coord: Coord,
  #[builder(into)]
  pub id: BuildingId,
  pub enabled: bool,
}
